zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -2788,40 +2803,124 @@ public void setACL(final String path, List<ACL> acl, int version, StatCallback c
+        final String serverPath = prependChroot(path);
 
         RequestHeader h = new RequestHeader();
         h.setType(ZooDefs.OpCode.getChildrenPaginated);
         GetChildrenPaginatedRequest request = new GetChildrenPaginatedRequest();
         request.setPath(serverPath);
         request.setWatch(watcher != null);
         request.setMaxReturned(maxReturned);
-        request.setMinCzxId(minCzxId);
-        request.setCzxIdOffset(czxIdOffset);
+        request.setMinCzxid(minCzxId);
+        request.setCzxidOffset(czxidOffset);
+
         GetChildrenPaginatedResponse response = new GetChildrenPaginatedResponse();
         ReplyHeader r = cnxn.submitRequest(h, request, response, wcb);
         if (r.getErr() != 0) {
             throw KeeperException.create(KeeperException.Code.get(r.getErr()),
-                    clientPath);
+                    path);
+        }
+        if (outputStat != null) {
+            DataTree.copyStat(response.getStat(), outputStat);
         }
+        updateNextPage(outputNextPage, response.getNextPageCzxid(), response.getNextPageCzxidOffset());
+
         return response.getChildren();
     }
 
+    /**
+     * Returns a list of all the children given the path.
+     * <p>
+     * The difference between this API and {@link #getChildren(String, boolean)} is:
+     * when there are lots of children and the network buffer exceeds {@code jute.maxbuffer},
+     * this API will fetch the children using pagination and be able to return all children;
+     * while {@link #getChildren(String, boolean)} will fail.
+     * <p>
+     * The final list of children returned is NOT strongly consistent with the server's data:
+     * the list might contain some deleted children if some children are deleted before the last page is fetched.
+     * <p>
+     * If the watch is true and the call is successful (no exception is thrown),
+     # a watch will be left on the node with the given path. The watch will be
+     * triggered by a successful operation that deletes the node of the given
+     * path or creates/deletes a child under the node.
+     *
+     * @param path the path of the parent node
+     * @param watch whether or not leave a watch on the given node
+     * @return a unordered list of all children of the given path
+     * @throws KeeperException if the server signals an error with a non-zero error code.
+     * @throws InterruptedException if the server transaction is interrupted.
+     */
+    public List<String> getAllChildrenPaginated(String path, boolean watch)
+            throws KeeperException, InterruptedException {
+        PaginationNextPage nextPage = new PaginationNextPage();
+        Watcher watcher = watch ? watchManager.defaultWatcher : null;
+        List<String> childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, 0L, 0, null, nextPage);
+        if (nextPage.getMinCzxid() == ZooDefs.GetChildrenPaginated.lastPageMinCzxid) {
```

Comment:
What is the static variable ZooDefs.GetChildrenPaginated.lastPageMinCzxid? Where do we update it? Does this mean we figure out whether we can return it in one page?

Reply:
It is a constant defined in https://github.com/linkedin/zookeeper/blob/7f606eac03b5c636aef2013eba0b3e389489054a/zookeeper-server/src/main/java/org/apache/zookeeper/ZooDefs.java#L164-L168

It represents the current fetched page is the final page, no more children left to fetch.

Reply:
I did not connect them together, how this ZooDefs.GetChildrenPaginated.lastPageMinCzxid get updated?

Reply:
@dasahcc it is a constant defined in  ZooDef which doesn't get changed.
The next page lastMinCzxid is updated in server when we calculate the next page info.